extensions/cl_ext_buffer_device_address.asciidoc

```diff
@@ -0,0 +1,315 @@
+
+Enums for enabling device pointer properties when creating a buffer
+{clCreateBuffer}, see <<clCreateBuffer, the list of supported memory flag values table>>:
+
+[source]
+----
+#define CL_MEM_DEVICE_ADDRESS_EXT (1ul << 31)
+#define CL_MEM_DEVICE_PRIVATE_EXT (1ul << 30)
+----
+
+Enums for querying the device pointer from the cl_mem <<clGetMemObjectInfo, the list of supported param_names table>>:
+
+[source]
+----
+#define CL_MEM_DEVICE_PTR_EXT 0xff01
+----
+
+Enums for setting information of indirect device pointer accesses to kernels <<clSetKernelExecInfo, the list of supported param_names table>>. This is for OpenCL 2.0 and above. When implementing the
+extension on an older OpenCL version, indirect device pointer access is not supported.
+
+[source]
+----
+#define CL_KERNEL_EXEC_INFO_DEVICE_PTRS_EXT 0x11B8
+----
+
+== New API Types
+
+Returned as the query result value *clGetMemObjectInfo* with `CL_DEVICE_PTR_EXT`.
+
+[source]
+----
+typedef cl_ulong cl_mem_device_address_EXT;
+----
+
+Returned as the query result value *clGetMemObjectInfo* with `CL_DEVICE_PTRS_EXT`.
+
+[source]
+----
+typedef struct _cl_mem_device_address_pair_EXT
+{
+  cl_device_id device;
+  cl_mem_device_address_EXT address;
+} cl_mem_device_address_pair_EXT;
+----
+
+== Modifications to the OpenCL API Specification
+
+=== Section 5.2.1 - Creating Buffer Objects:
+
+Add new allocation flags <<clCreateBuffer, List of supported memory flag values table>>:
+
+[[list-of-supported-memory-flag-values-adds]]
+.List of supported memory flags by {clCreateBuffer}
+[width="100%",cols="<50%,<50%",options="header"]
+|====
+| Memory Flags | Description
+| {CL_MEM_DEVICE_ADDRESS_EXT_anchor}
+
+include::{generated}/api/version-notes/CL_MEM_DEVICE_ADDRESS_EXT.asciidoc[]
+  | This flag specifies that the buffer must have a single fixed address
+   for its lifetime and the address should be unique at least across the devices
+   of the context, but not necessarily within the host (virtual) memory.
+
+   The flag might imply that the buffer will be "pinned" permanently to
+   a device's memory, but might not be necessarily so, as long as the address
+   range of the buffer remains constant.
+
+   The address is guaranteed to remain the same until the buffer is freed, and
+   the address can be queried via {clGetMemObjectInfo}.
+
+   The device-specific buffer content updates are still performed by
+   implicit or explicit buffer migrations performed by the runtime or the
+   client code. If all of the devices in the context do not support
+   this type of allocations, an error (CL_INVALID_VALUE) is returned.
+
+   The device addresses of sub-buffers derived from CL_MEM_DEVICE_ADDRESS_EXT
+   allocated buffers can be computed by adding the sub-buffer origin to the
+   start address.
+
+| {CL_MEM_DEVICE_PRIVATE_EXT_anchor}
+
+include::{generated}/api/version-notes/CL_MEM_DEVICE_PRIVATE_EXT.asciidoc[]
+  | If this flag is combined with CL_MEM_DEVICE_ADDRESS_EXT, each device in
```

Comment:
I think it makes more sense to remove the `combined with CL_MEM_DEVICE_ADDRESS_EXT` part, then those two flags are independent.

Then a `CL_DEVICE_SUPPORTS_MULTI_DEVICE_ADDRESS` query could be added to `clGetDeviceInfo` to indicate if `CL_MEM_DEVICE_ADDRESS_EXT` is supported when the device is used in a context with other devices.

This should be enough to make this extension supportable on vulkan as mentioned earlier.